### PR TITLE
Vue 3

### PR DESCRIPTION
**Pull request by raimund-schluessler**
This PR aims at migrating all components to vue3.

Known issues:
- [x] Rendering markdown with `NcRichText` does not work
- [ ] There is no vue3 compatible release of `@nextcloud/vue-select` yet

Breaking changes:
- Requires the app to use vue3
- The `exact` prop of `NcAppNavigationItem`, `NcBreadcrumb`, `NcButton` and `NcListItem` components is removed as vue-router 4 does not support it anymore

Compiling the Tasks apps [vue3](https://github.com/nextcloud/tasks/tree/vue3) branch from https://github.com/nextcloud/tasks/pull/1971 together with this PR here already gives a fully working app. See here:
Screenshot 2023-03-07 at 12-59-11 Aufgaben - Nextcloud

### CODE REVIEW

**Comment by raimund-schluessler** [comment]
@skjnldsv Do you maybe have an idea why `npm run styleguide:build` fails with this error (`npm run styleguide` works though):
```js
4:55:09 PM: Building @nextcloud/vue 7.5.0
4:55:09 PM: This build version hash is 69d54a5
4:55:16 PM: SyntaxError: Unexpected token (1:5)
4:55:16 PM: SyntaxError: Unexpected token (1:5)
4:55:16 PM:     at Object.pp$4.raise (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:2825:15)
4:55:16 PM:     at Object.pp.unexpected (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:689:10)
4:55:16 PM:     at Object.pp.semicolon (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:666:66)
4:55:16 PM:     at Object.pp$1.parseExpressionStatement (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:1145:10)
4:55:16 PM:     at Object.pp$1.parseStatement (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:880:26)
4:55:16 PM:     at Object.pp$1.parseTopLevel (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:746:23)
4:55:16 PM:     at Object.parse (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:553:17)
4:55:16 PM:     at Function.parse (/opt/build/repo/node_modules/vue-inbrowser-compiler/node_modules/acorn/dist/acorn.js:576:37)
4:55:16 PM:     at getAst (/opt/build/repo/node_modules/vue-inbrowser-compiler/lib/vue-inbrowser-compiler.cjs.js:19:27)
4:55:16 PM:     at prepareVueCodeForEvalFunction (/opt/build/repo/node_modules/vue-inbrowser-compiler/lib/vue-inbrowser-compiler.cjs.js:260:17)
4:55:16 PM:     at Object.compileVueCodeForEvalFunction [as compile] (/opt/build/repo/node_modules/vue-inbrowser-compiler/lib/vue-inbrowser-compiler.cjs.js:225:34)
4:55:16 PM:     at /opt/build/repo/node_modules/vue-styleguidist/lib/loaders/examples-loader.js:180:52
4:55:16 PM:     at Array.map (<anonymous>)
4:55:16 PM:     at Object._callee$ (/opt/build/repo/node_modules/vue-styleguidist/lib/loaders/examples-loader.js:167:41)
4:55:16 PM:     at tryCatch (/opt/build/repo/node_modules/vue-styleguidist/node_modules/@babel/runtime/helpers/regeneratorRuntime.js:44:17)
4:55:16 PM: Failed during stage 'building site': Build script returned non-zero exit code: 2 (https://ntl.fyi/exit-code-2)
4:55:16 PM:     at Generator.<anonymous> (/opt/build/repo/node_modules/vue-styleguidist/node_modules/@babel/runtime/helpers/regeneratorRuntime.js:125:22)
4:55:16 PM:     at Generator.next (/opt/build/repo/node_modules/vue-styleguidist/node_modules/@babel/runtime/helpers/regeneratorRuntime.js:69:21)
4:55:16 PM:     at asyncGeneratorStep (/opt/build/repo/node_modules/vue-styleguidist/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
4:55:16 PM:     at _next (/opt/build/repo/node_modules/vue-styleguidist/node_modules/@babel/runtime/helpers/asyncToGenerator.js:22:9)
4:55:16 PM:     at /opt/build/repo/node_modules/vue-styleguidist/node_modules/@babel/runtime/helpers/asyncToGenerator.js:27:7
4:55:16 PM:     at new Promise (<anonymous>)
4:55:16 PM:     at Object.<anonymous> (/opt/build/repo/node_modules/vue-styleguidist/node_modules/@babel/runtime/helpers/asyncToGenerator.js:19:12)
4:55:16 PM:     at Object.examplesLoader (/opt/build/repo/node_modules/vue-styleguidist/lib/loaders/examples-loader.js:56:26)
4:55:16 PM:     at Object._default (/opt/build/repo/node_modules/vue-styleguidist/lib/loaders/examples-loader.js:51:18)
4:55:16 PM:     at LOADER_EXECUTION (/opt/build/repo/node_modules/loader-runner/lib/LoaderRunner.js:132:14)
4:55:16 PM:     at runSyncOrAsync (/opt/build/repo/node_modules/loader-runner/lib/LoaderRunner.js:133:4)
4:55:16 PM:     at iterateNormalLoaders (/opt/build/repo/node_modules/loader-runner/lib/LoaderRunner.js:251:2)
4:55:16 PM:     at /opt/build/repo/node_modules/loader-runner/lib/LoaderRunner.js:224:4
4:55:16 PM:     at /opt/build/repo/node_modules/webpack/lib/NormalModule.js:834:15
4:55:16 PM:     at eval (eval at create (/opt/build/repo/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:12:1)
4:55:16 PM:     at processTicksAndRejections (node:internal/process/task_queues:83:21)
4:55:16 PM:     at runNextTicks (node:internal/process/task_queues:65:3)
4:55:16 PM:     at processImmediate (node:internal/timers:437:9)
4:55:16 PM: ​
4:55:16 PM:   "build.command" failed                                        
4:55:16 PM: ────────────────────────────────────────────────────────────────
```

**Comment by skjnldsv** [comment]
> @skjnldsv Do you maybe have an idea why `npm run styleguide:build` fails with this error (`npm run styleguide` works though):

Nope :(

**Comment by raimund-schluessler** [comment]
Blocked by #3775 for now.

**Comment by ShGKme** [comment]
> Requires the app to use vue3

With [vue-demi](https://github.com/vueuse/vue-demi), it is possible to make the library compatible with both Vue 2 and Vue 3.

It could be tricky in some places where the components are very different in Vue 2 and Vue 3 (for example, render function). But it would make it possible not to support 2 different versions of `nextcloud-vue` and avoid a big breaking change.

**Comment by raimund-schluessler** [comment]
> > Requires the app to use vue3
> 
> With [vue-demi](https://github.com/vueuse/vue-demi), it is possible to make the library compatible with both Vue 2 and Vue 3.

Honestly, I would prefer a clear cut. Vue 2 will reach [end of life](https://v2.vuejs.org/lts/) by December 31st, 2023. I don't see `@nextcloud/vue` support vue 3 much earlier, so why spend this extra effort just to support vue 2 for a couple of months.

**Comment by ShGKme** [comment]
> Honestly, I would prefer a clear cut. Vue 2 will reach [end of life](https://v2.vuejs.org/lts/) by December 31st, 2023. I don't see `@nextcloud/vue` support vue 3 much earlier, so why spend this extra effort just to support vue 2 for a couple of months.

Do you mean that we want to migrate all the apps to Vue 3 soon, making `@nextcloud/vue` for Vue 2 deprecated?

**Comment by raimund-schluessler** [comment]
> Do you mean that we want to migrate all the apps to Vue 3 soon, making `@nextcloud/vue` for Vue 2 deprecated?

I think that's a question for the Nextcloud company, and I am just a contributor. Unfortunately, they are a bit silent regarding this question, though.

Having a stable vue-3-only release does not necessarily mean not developing fixes or even features for the vue-2-release. One can always backport code. I just think that putting effort in a release that supports both 2 and 3 is not so well spent, given that we *have* to migrate rather sooner than later if we want to keep our important dependencies supported.

**Comment by susnux** [comment]
I think migrating all apps within one release is quite unrealistic.
But nevertheless I think the best way would be to focus on vue 3 only, as Raimund already said vue 2 is about to be EOL and vue 2 compatibility would introduce much more complexity and require more manpower.

This way we could maybe release a vue3 version (maybe 8.0?) and keep maintaining version 7.x for vue 2 until most apps are migrated. For example it would be nice to have a beta version for testing with NC27.
(Nextcloud versions are just examples)

**Comment by raimund-schluessler** [comment]
Hooray, with the latest release of vue-styleguidist, `npm run styleguide:build` works properly. We now have a working vue3 version of the documentation :tada: :tada: 

Time to merge this PR, I guess :laughing: :wink:  :see_no_evil:

**Comment by dartcafe** [comment]
Just a thought: Would it be a good idea to fork this project into a new repo (i.e. nextcloud-vue3).

**Comment by raimund-schluessler** [comment]
> Just a thought: Would it be a good idea to fork this project into a new repo (i.e. nextcloud-vue3). 

Hm, that would make it quite difficult to keep track of changes and improvements from nextcloud-vue.

**Comment by dartcafe** [comment]
> Hm, that would make it quite difficult to keep track of changes and improvements from nextcloud-vue.

Or nextcloud-vue changes to maintenance only and nextcloud-vue3 becomes the new main branch. In the end vue2 itselfs approches eol, so should nextcloud-vue too.

BTW: How do I make dependabot walk through other branches than the main branch?

**Comment by susnux** [comment]
Maybe take the opportunity and use in version 8 starting with the NC28 cycle? :)

**Comment by raimund-schluessler** [comment]
> Maybe take the opportunity and use in version 8 starting with the NC28 cycle? :)

It would be nice if we would get a beta release in the next time, but making it default for NC28 might be to ambitious (although I would like it :slightly_smiling_face:).

**Comment by susnux** [comment]
Sorry it was already late when writing that, what I meant was:
Having the nextcloud-vue 8 release as a breaking release would allow also for beta testing the vue 3 version as it is already a breaking change. Of cause we should test it properly before making it the default :)

**Comment by susnux** [comment]
@raimund-schluessler 

> The exact prop of NcAppNavigationItem, NcBreadcrumb, NcButton and NcListItem components is removed as vue-router 4 does not support it anymore

Maybe we should already deprecate this properties for v8 of `@nextcloud/vue`?

**Comment by raimund-schluessler** [comment]
> @raimund-schluessler
> 
> > The exact prop of NcAppNavigationItem, NcBreadcrumb, NcButton and NcListItem components is removed as vue-router 4 does not support it anymore
> 
> Maybe we should already deprecate this properties for v8 of `@nextcloud/vue`?

I am unsure if this makes sense, since there is no way to fix the deprecation message before really switching to vue3. I would say deprecating is only helpful if you can already migrate to the alternative.

**Review by ShGKme — COMMENTED** [review]
Hi Raimund!

I've started to look into this migration. I haven't proper check every component yet, still, there are some things that seems to be broken or can be improved for me.

File: src/components/NcMultiselect/NcMultiselect.vue
[the remark this answers is not in the document]
Comment (raimund-schluessler) [reply src/components/NcMultiselect/NcMultiselect.vue:1]: Yes, needs to be removed, and was removed in master already.

File: src/components/NcVNodes/NcVNodes.vue
[the remark this answers is not in the document]
Comment (raimund-schluessler) [reply src/components/NcVNodes/NcVNodes.vue:1]: If that works, lets remove the component.

File: src/components/NcAppSidebar/NcAppSidebar.vue [hunk 1]
```diff
@@ -626,7 +627,7 @@ export default {
 		'opened',
 		'figure-click',
 		'update:starred',
-		'update:titleEditable',
+		'update:title-editable',
```
[the remark this answers is not in the document]
Comment (raimund-schluessler) [reply src/components/NcAppSidebar/NcAppSidebar.vue:630]: I think we switched to kebab case at any other place, but I am not sure. Would need to be checked.

File: src/components/NcAppNavigation/NcAppNavigation.vue [hunk 2]
```diff
@@ -90,6 +90,9 @@ export default {
 	data() {
 		return {
 			open: true,
+			// Watching an injected property directly does not work
+			// so we watch a data property instead.
+			localIsMobile: this.isMobile,
```
[the remark this answers is not in the document]
Comment (raimund-schluessler) [reply src/components/NcAppNavigation/NcAppNavigation.vue:95]: Yes, the last time I touched this branch Vue 3.3 was not release. Good that it's there now.

**Comment by raimund-schluessler** [comment]
> Hi Raimund!
> 
> I've started to look into this migration. I haven't proper check every component yet, still, there are some things that seems to be broken or can be improved for me.

Yes, the last time I touched vue `vue3` branch was before many breaking changes got merged into master and before the release of Vue 3.3. When rebasing `vue3` onto master, many things would need to be updated and adjusted now.

I don't know how much time I will have for this in the near future. If you like to go ahead, feel free to do so.

**Comment by raimund-schluessler** [comment]
I rebased locally, but cannot currently push to the repo. Seems I don't have write access.

File: src/components/NcMultiselect/NcMultiselect.vue
[the remark this answers is not in the document]
Comment (raimund-schluessler) [reply src/components/NcMultiselect/NcMultiselect.vue:1]: Done.

**Comment by raimund-schluessler** [comment]
@ShGKme I pushed my rebase. It includes all changes to master, and I fixed some of your comment, but not all yet. Mainly because the styleguide currently does not build, and I couldn't really test whether the changes would work. I would propose to first get the styleguide and the vite build working in this branch `vue3` and then start creating PRs for the components one by one to a `next` branch as you suggested.

@susnux I couldn't get the vite build and the styleguide to work, probably, because I don't know much about vite and typescript. If you would have some time to have a look at this, it would be great. Also the plugin test does not work yet, probably for the same reason.

File: src/components/NcAppSidebar/NcAppSidebar.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShGKme) [reply src/components/NcAppSidebar/NcAppSidebar.vue]: It was recommended in Vue 2 to use `kebab-case`, but this recommendation was changed

[the remark this answers is not in the document]
Comment (raimund-schluessler) [reply src/components/NcAppSidebar/NcAppSidebar.vue]: I am fine with whatever case we agree on, I would just like to keep it consistent (and ideally have the linter configured to complain about the wrong case).

**Comment by raimund-schluessler** [comment]
@ShGKme @skjnldsv Thanks to @susnux the configuration and infrastructure now looks good. The styleguide works, the tests run (and the majority passes, just two `NcRichStuff` related ones fail) and most components work fine in the styleguide.

How should we proceed now? Create a `next` branch from `master` and then create PRs little by little with a first PR e.g. containing the configuration and a simple component with the rest commented out?

**Comment by susnux** [comment]
@raimund-schluessler 
> I couldn't get the vite build and the styleguide to work. Probably, because I don't know much about vite and typescript. If you would have some time to have a look at this, it would be great. Also the plugin.spec.ts test still fails, probably for the same reason.

I fixed the vite config (there is now a vue3 branch of it) and fixed the plugin test as vue3 changed the way components are resolved.

**Comment by susnux** [comment]
*(why was my answer above sent with such an delay?)*

> How should we proceed now? Create a next branch from master and then create PRs little by little with a first PR e.g. containing the configuration and a simple component with the rest commented out?

That sounds like the best way to go :)

**Comment by ShGKme** [comment]
> How should we proceed now? Create a `next` branch from `master` and then create PRs little by little with a first PR e.g. containing the configuration and a simple component with the rest commented out?

Yeah, let's go this way!

1. Create the `next` branch (which will be `v9.0` then I hope :D)
2. Merge infrastructure updates with a commented list of the components in styleguidist and 1-2 components as an example
3. Then add individual components in new PRs

And then backport all changes in PRs not only to `stable7`, but also to `next`.

@skjnldsv what do you think?

**Comment by susnux** [comment]
We should also discuss some other things when doing this major migration:

* Discuss if we want to move all components to templates instead of render functions. Because at least with vue3 the code from the template compiler is mostly better optimized than custom render functions.
* Add type definitions to props and emits so we can provide full Typescript support.
* Do we want to stick with Options-API or do we want to migrate to composition API¹

Of cause this is not needed to be done right away, but can be done step-by-step afterwards.

But I do not want to snub anyone so if its just me preferring those things, than we should stick with the current style as you did a great job with this library and its current state :)

---

¹I throw in the composition API for discussion as usables are from my point of view cleaner than mixins, especially when mixins rely on something defined in a **peer* mixin (and even as mixins are order dependent).
(Also that is the recommended API by Vue, and most documentation and courses are about it, so probably future contributors would be more familar with it)

**Comment by raimund-schluessler** [comment]
>     * Discuss if we want to move all components to templates instead of render functions. Because at least with vue3 the code from the template compiler is mostly better optimized than custom render functions.

I was under the impression that some things cannot properly be achieved by using templates, hence, render functions would be necessary sometimes. E.g. for the `NcActions` component. But if the same works with templates fine with me. However, I would prefer to first move existing render functions to vue3, and move to templates afterwards, since I assume that simply migrating them might be easier.

>     * Add type definitions to props and emits so we can provide full Typescript support.

Sure, why not. Although I have no experience with Typescript yet, I have to admit.
 
>     * Do we want to stick with Options-API or do we want to migrate to composition API¹

I don't have experience with the composition API. However, migrating to composition should be non-breaking, I guess, so it even can be done after a first vue3 release.

**Comment by skjnldsv** [comment]
> How should we proceed now? Create a `next` branch from `master` and then create PRs little by little with a first PR e.g. containing the configuration and a simple component with the rest commented out?

I guess so yes :/
Migration will be cumbersome

**Comment by raimund-schluessler** [comment]
The components are now fully migrated, so this PR is not needed anymore.

**Pull request closed without merging**